transport-native-epoll/src/main/c/netty_epoll_native.c

```diff
@@ -169,42 +168,37 @@ static jint netty_epoll_native_epollCreate(JNIEnv* env, jclass clazz) {
     return efd;
 }
 
-static jint netty_epoll_native_epollWait0(JNIEnv* env, jclass clazz, jint efd, jlong address, jint len, jint timeout) {
+static jint netty_epoll_native_epollWait0(JNIEnv* env, jclass clazz, jint efd, jlong address, jint len, jint timerFd, jint tvSec, jint tvNsec) {
     struct epoll_event *ev = (struct epoll_event*) (intptr_t) address;
-    struct timespec ts;
-    jlong timeBeforeWait, timeNow;
-    int timeDiff, result, err;
-
-    if (timeout > MAX_EPOLL_TIMEOUT_MSEC) {
-        // Workaround for bug in older linux kernels that can not handle bigger timeout then MAX_EPOLL_TIMEOUT_MSEC.
-        timeout = MAX_EPOLL_TIMEOUT_MSEC;
-    }
-
-    clock_gettime(waitClockId, &ts);
-    timeBeforeWait = ts.tv_sec * 1000 + ts.tv_nsec / 1000000;
-
-    for (;;) {
-      result = epoll_wait(efd, ev, len, timeout);
-      if (result >= 0) {
-        return result;
-      }
-      if ((err = errno) == EINTR) {
-        if (timeout > 0) {
-          clock_gettime(waitClockId, &ts);
-          timeNow = ts.tv_sec * 1000 + ts.tv_nsec / 1000000;
-          timeDiff = timeNow - timeBeforeWait;
-          timeout -= timeDiff;
-          if (timeDiff < 0 || timeout <= 0) {
-            return 0;
-          }
-          timeBeforeWait = timeNow;
-        } else if (timeout == 0) {
-          return 0;
+    int result, err;
+
+    if (tvSec == 0 && tvNsec == 0) {
+        // Zeros = poll (aka return immediately).
+        do {
+            result = epoll_wait(efd, ev, len, 0);
+            if (result >= 0) {
+              return result;
+            }
+        } while((err = errno) == EINTR);
+    } else {
+        struct itimerspec ts;
+        memset(&ts.it_interval, 0, sizeof(struct timespec));
+        ts.it_value.tv_sec = tvSec;
+        ts.it_value.tv_nsec = tvNsec;
+        result = timerfd_settime(timerFd, 0, &ts, NULL);
```

Comment:
agreed. I will throw together some perf tests, but it would be nice to have some more real world evaluation. Perhaps someone involved in https://issues.apache.org/jira/browse/CASSANDRA-13651 can run some tests ... can you reach out?